### DISCUSSION TITLE
SQL plugin install

### DISCUSSION THREAD

**Discussion by Blatman** [tauri-apps/tauri#9254] [Q&A] [answered]
Just want to clarify a bit -
I installed the sql plugin and works fine however I only had to add a line to the Cargo.toml and it appeared to do a whole lot of cargo upgrade stuff.

tauri-plugin-sql = { version = "2.0.0-beta.3", features = ["sqlite"] }

The docs suggest that the JS bindings are installed with npm/yarn ( I use npm) but I didn't do this and it still works - it looks like the .toml setting installs the plugin to the .cargo directories and is used from there. 
So is it optional to install via cargo or use npm install rather than needing both?

If both were done would this create a potential version conflict issue or would one install take precedence?

<img width="779" alt="installdoc" src="https://github.com/tauri-apps/tauri/assets/11089335/ac6a7af2-9f8b-4ef3-b183-5612b3020c39">

**Comment by FabianLars** [comment]
The cargo/rust installation is required to make tauri plugins work, whether or not you use its rust apis. The npm package is kinda optional, only needed if you want to use the plugin's js apis (though plugin apis are typically also available via `window.__TAURI__`.

**Reply by Blatman** [reply]
Ah OK that answers why the JS side is working without the npm install in @tauri-apps modules folder because I use the window.__TAURI__ method to access the API. I guess that means the window.__TAURI__.sql method hooks into the cargo install but would hook into the npm install if it had been done.

**Reply by FabianLars** [reply]
> I guess that means the window.TAURI.sql method hooks into the cargo install but would hook into the npm install if it had been done.

No. The `window__TAURI__` api and the npm package are basically the same thing, just 2 different distribution systems.

The rust package is required either way! All plugin js apis (no matter if sql or any of the others) are basically just prebuilt [tauri commands](https://tauri.app/v1/guides/features/command/) which are always writting in rust.

**Reply by Blatman** [reply]
Mmmm - OK the plugins are using the rust so a bit mystified why the npm install would be needed at all. Most of the other plugins all appear in the @tauri-apps folder because I used an npm install but the sql one doesn't as only the toml addition was done but still works with window.TAURI.sql method (I can never get the standard import method to work, get path errors) - anyway thanks for clarifying that the cargo install is the necessary one.

**Reply by FabianLars** [reply]
> I can never get the standard import method to work, get path errors

Then i assume you're not using a bundler like vite or webpack? A bundler is basically a requirement to use npm packages. If you do use one and still see errors feel free to share them, maybe we messed something up.

> a bit mystified why the npm install would be needed at all.

To many devs importing from npm modules feels just cleaner/better (edit: probably because tooling is typically built for this more than using the window var). But aside from feelings, it will result in lower binary size because the bundler can strip out unused code and/or minimize it.
Edit2: Also, the npm package offers typings. Once you use typescript you'll have a hard time making it happy with the window var apis.

**Reply by Blatman** [reply, marked as answer]
Ah thankyou that explains it - I tend not to use a bundler only what Tauri provides in it's build mechanism.
